### PR TITLE
[TT-1570/TT-12587]mTLS: allow validating client certificates against root certificate authority

### PR DESCRIPTION
**Pull request by jeffy-mathew**
## Description
mTLS: allow validating client certificates against root certificate authority
## Related Issue
Parent: https://tyktech.atlassian.net/browse/TT-1570
Subtask: https://tyktech.atlassian.net/browse/TT-12587
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added comprehensive tests for validating client certificates against a root certificate authority in mutual TLS scenarios.
- Implemented helper functions to generate root, server, and client certificates and keys for testing purposes.
- Enhanced the `ValidateRequestCerts` function to handle multiple peer certificates and improved error handling.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>cert_test.go</strong><dd><code>Add mTLS validation tests and certificate generation helpers</code></dd></summary>
<hr>

gateway/cert_test.go

<li>Added tests for validating client certificates against a root <br>certificate authority.<br> <li> Implemented helper functions to generate root, server, and client <br>certificates and keys.<br> <li> Enhanced mutual TLS tests to include valid and invalid client <br>scenarios.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6405/files#diff-481696cb18de8c3880e0ca21318fe00fd4eb89bc48994e43b7c34729ef4a7ee2">+328/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>helpers.go</strong><dd><code>Enhance certificate validation logic in mTLS</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/crypto/helpers.go

<li>Modified <code>ValidateRequestCerts</code> to handle multiple peer certificates.<br> <li> Improved error handling for certificate validation.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6405/files#diff-3d1fc755c46eaa99e9f1edf358b9e00842342ae6333902959d7a68b46d156829">+22/-17</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: gateway/cert_test.go [hunk 1]
```diff
@@ -1732,4 +1807,257 @@ func TestStaticMTLSAPI(t *testing.T) {
...
+	if rootCertBlock == nil || rootCertBlock.Type != "CERTIFICATE" {
+		return nil, nil, fmt.Errorf("failed to decode root certificate PEM")
+	}
+	rootCert, err := x509.ParseCertificate(rootCertBlock.Bytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Decode the root private key
+	rootKeyBlock, _ := pem.Decode(rootKeyPEM)
```
Comment (titpetric) [inline gateway/cert_test.go:1934]: The omitted values are errors? Cover with assert.NoError(), or add error checks for swallowed values

File: gateway/cert_test.go [hunk 2]
```diff
@@ -1732,4 +1807,257 @@ func TestStaticMTLSAPI(t *testing.T) {
...
+			req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("localhost:%s/static-mtls", u.Port()), nil)
+			assert.NoError(t, err)
+			_, err = invalidClient.Do(req)
+			assert.Error(t, err)
+		})
+
+	})
+}
+
+func generateRootCertAndKey(tb testing.TB) ([]byte, []byte, error) {
```
Comment (titpetric) [inline gateway/cert_test.go:1879]: Move to internal/crypto/testutil.go :) for all the generate funcs. Cover individually with some unit tests?

File: gateway/cert_test.go [hunk 3]
```diff
@@ -1732,4 +1807,257 @@ func TestStaticMTLSAPI(t *testing.T) {
...
+			tlsConfig.InsecureSkipVerify = false
+			transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+			invalidClient := &http.Client{Transport: transport}
+			u, err := url.Parse(ts.URL)
+
+			req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("localhost:%s/static-mtls", u.Port()), nil)
+			assert.NoError(t, err)
+			_, err = invalidClient.Do(req)
+			assert.Error(t, err)
```
Comment (titpetric) [inline gateway/cert_test.go:1873]: We may need to have a stronger assertion for the error returned here, so we know it's related to TLS. I asume the request could fail due to some routing related issue (invalid host, rate limit, etc.).

File: gateway/cert_test.go [hunk 4]
```diff
@@ -1732,4 +1807,257 @@ func TestStaticMTLSAPI(t *testing.T) {
...
+		t.Run("invalid client", func(t *testing.T) {
+			_, _, _, invalidClientCert := crypto.GenCertificate(&x509.Certificate{}, false)
+			tlsConfig := GetTLSConfig(&invalidClientCert, nil)
+			tlsConfig.InsecureSkipVerify = false
+			transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+			invalidClient := &http.Client{Transport: transport}
+			u, err := url.Parse(ts.URL)
+
+			req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("localhost:%s/static-mtls", u.Port()), nil)
```
Comment (titpetric) [inline gateway/cert_test.go:1870]: Can you just use `ts.URL` here?

Comment (jeffy-mathew) [reply gateway/cert_test.go:1870]: no, we need to use domain as `localhost`

File: gateway/cert_test.go (hunk 2 above)
Comment (jeffy-mathew) [reply gateway/cert_test.go:1879]: We could move it to testutil.go, but covering it with another set of unit tests seem to be an overkill imo.

File: gateway/cert_test.go (hunk 1 above)
Comment (jeffy-mathew) [reply gateway/cert_test.go:1934]: It's not error, it's the `rest`/`remainder` after decoding PEM block

File: .github/workflows/ci-tests.yml [hunk 5]
```diff
@@ -135,7 +135,7 @@ jobs:
             -Dsonar.projectKey=TykTechnologies_tyk
             -Dsonar.sources=.
             -Dsonar.exclusions=**/testdata/*,test/**,coprocess/**/*,ci/**,smoke-tests/**,apidef/oas/schema/schema.gen.go
-            -Dsonar.coverage.exclusions=**/*_test.go,**/mock/*
+            -Dsonar.coverage.exclusions=**/*_test.go,**/mock/*,internal/crypto/testutil.go
```
Comment (titpetric) [inline .github/workflows/ci-tests.yml:138]: Revert this, include it in coverage

File: gateway/cert_test.go (hunk 3 above)
Comment (titpetric) [reply gateway/cert_test.go:1869]: @jeffy-mathew unresolved

File: internal/crypto/testutil.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+	// Create the root certificate
+	rootCertDER, err := x509.CreateCertificate(rand.Reader, &rootCertTemplate, &rootCertTemplate, &rootKey.PublicKey, rootKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Encode the root certificate to PEM format
+	var rootCertPEM bytes.Buffer
+	_ = pem.Encode(&rootCertPEM, &pem.Block{Type: "CERTIFICATE", Bytes: rootCertDER})
```
Comment (titpetric) [inline internal/crypto/testutil.go:62]: These two should have assert.NoError(tb, <x>) wrapped

File: .github/workflows/ci-tests.yml (hunk 5 above)
Comment (jeffy-mathew) [reply .github/workflows/ci-tests.yml:138]: It's a test helper and since it's a separate package it wouldn't be reported in the coverage

Comment (titpetric) [reply .github/workflows/ci-tests.yml:138]: let it be reported, the package is /internal and it's an importable API, let's consider coverage confirms usage

**Review by titpetric — APPROVED** [review]
Revert ci-tests.yml and you're good

**Pull request merged**